src/replication.c

```diff
@@ -625,59 +639,88 @@ typedef struct read_event {
 	int *read_count;
 } read_event_t;
 
-int read_io_resp(spec_t *spec, replica_t *replica, read_event_t *revent);
+int read_and_process_io_resp(spec_t *spec, replica_t *replica, read_event_t *revent);
 
+replica_t *
+get_replica_from_iofd(spec_t *spec, int iofd) {
+	replica_t *replica;
+	MTX_LOCK(&spec->rq_mtx);
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if(iofd == replica->iofd) {
+			MTX_LOCK(&replica->r_mtx);
+			__sync_fetch_and_add(&replica->refcount, 1);
+			MTX_UNLOCK(&replica->r_mtx);
+			MTX_UNLOCK(&spec->rq_mtx);
+			return replica;
+		}
+	}
+	MTX_UNLOCK(&spec->rq_mtx);
+	return NULL;
+}
 //Receive IO responses in this thread
 void *
 replica_receiver(void *arg) {
 	spec_t *spec = (spec_t *)arg;
 	replica_t *replica;
-	int i, event_count, replica_count = 0, ret;
+	int i, event_count, ret;
 	struct epoll_event event, *events;
 	read_event_t revent;
 
 	//Create a new epoll epfd
 	int epfd = spec->receiver_epfd;
 	events = calloc(MAXEVENTS, sizeof(event));
 	while(1) {
-		if(replica_count == 0) {
-			MTX_LOCK(&spec->rq_mtx);
-			if(spec->replica_count == 0) {
-				//Wait until at least one IO connection has been made to a registered replica
-				pthread_cond_wait(&spec->rq_cond, &spec->rq_mtx);
-			}
-			MTX_UNLOCK(&spec->rq_mtx);
-		}
 		//Wait for events on all iofds
 		event_count = epoll_wait(epfd, events, MAXEVENTS, -1);
 		for(i=0; i< event_count; i++) {
 			//Remove the replica from queue
 			if ((events[i].events & EPOLLERR) ||
 					(events[i].events & EPOLLHUP) ||
 					(!(events[i].events & EPOLLIN))) {
-				remove_replica_from_list(spec, events[i].data.fd);
+				get_replica_from_iofd(spec, events[i].data.fd);
+				remove_replica_from_list(spec, replica);
 				REPLICA_LOG("epoll error\n");
 				continue;
 			} else {
 				MTX_LOCK(&spec->rq_mtx);
 				TAILQ_FOREACH(replica, &spec->rq, r_next) {
 					if(events[i].data.fd == replica->iofd) {
+						MTX_LOCK(&replica->r_mtx);
+						if(replica->state == REPLICA_ERRORED) {
```

Comment:
this check is not required if we make sure that we remove from list before changing state to ERRORED..
so, this need to be removed based on changes in remove_replica_from_list